libcloud/storage/drivers/digitalocean_spaces.py

```diff
@@ -0,0 +1,101 @@
+
+__all__ = [
+    'DigitalOceanSpacesStorageDriver'
+]
+
+DO_SPACES_HOSTS_BY_REGION = {'nyc3': 'nyc3.digitaloceanspaces.com'}
+
+DO_SPACES_DEFAULT_REGION = 'nyc3'
+
+
+class DOSpacesConnectionAWS4(SignedAWSConnection, BaseS3Connection):
+    service_name = 's3'
+    version = API_VERSION
+
+    def __init__(self, user_id, key, secure=True, host=None, port=None,
+                 url=None, timeout=None, proxy_url=None, token=None,
+                 retry_delay=None, backoff=None, **kwargs):
+
+        super(DOSpacesConnectionAWS4, self).__init__(user_id, key,
+                                                     secure, host,
+                                                     port, url,
+                                                     timeout,
+                                                     proxy_url, token,
+                                                     retry_delay,
+                                                     backoff,
+                                                     4)  # force aws4
+
+
+class DOSpacesConnectionAWS2(S3Connection):
+
+    def __init__(self, user_id, key, secure=True, host=None, port=None,
+                 url=None, timeout=None, proxy_url=None, token=None,
+                 retry_delay=None, backoff=None, **kwargs):
+
+        super(DOSpacesConnectionAWS2, self).__init__(user_id, key,
+                                                     secure, host,
+                                                     port, url,
+                                                     timeout,
+                                                     proxy_url, token,
+                                                     retry_delay,
+                                                     backoff)
+
+
+class DigitalOceanSpacesStorageDriver(S3StorageDriver):
+    name = 'DigitalOcean Spaces'
+    website = 'https://www.digitalocean.com/products/object-storage/'
+    supports_chunked_encoding = False
+    supports_s3_multipart_upload = True
+
+    def __init__(self, key, secret=None, secure=True, host=None, port=None,
+                 api_version=None, region=DO_SPACES_DEFAULT_REGION, **kwargs):
+
+        if region not in DO_SPACES_HOSTS_BY_REGION:
+            raise LibcloudError('Unknown region (%s)' % (region), driver=self)
+
+        host = DO_SPACES_HOSTS_BY_REGION[region]
+        self.name = 'DigitalOcean Spaces (%s)' % (region)
+
+        self.region_name = region
+        self.signature_version = str(kwargs.pop('signature_version',
+                                                DEFAULT_SIGNATURE_VERSION))
+
+        if self.signature_version not in ['2', '4']:
+            raise ValueError('Invalid signature_version: %s' %
+                             (self.signature_version))
+
+        if self.signature_version == '2':
+            self.connectionCls = DOSpacesConnectionAWS2
+        elif self.signature_version == '4':
+            self.connectionCls = DOSpacesConnectionAWS4
```

Comment:
nitpick: Putting the ValueError exception as an `else` branch would avoid repeating '2' and '4'.